alpha/veneer/basic/basic.go

```diff
@@ -0,0 +1,101 @@
+package basic
+
+import (
+	"context"
+	"errors"
+	"os"
+	"path/filepath"
+
+	"github.com/operator-framework/operator-registry/alpha/action"
+	"github.com/operator-framework/operator-registry/alpha/declcfg"
+	"github.com/operator-framework/operator-registry/alpha/property"
+	"github.com/operator-framework/operator-registry/pkg/image"
+)
+
+var ErrNotAllowed = errors.New("not allowed")
+var ErrCannotIngestDirectory = errors.New("cannot ingest veneers by directory reference")
+
+type BasicVeneer struct {
+	Registry image.Registry
+}
+
+func (v BasicVeneer) Ingest(ctx context.Context, ref string) (*declcfg.DeclarativeConfig, error) {
+	// only taking first argument as file
+	stat, serr := os.Stat(ref)
+	if serr == nil {
+		if stat.IsDir() {
+			return nil, ErrCannotIngestDirectory // BUGBUG enhance to walk and eat all files in turn, heeding ignore file
+		} else {
+			return v.renderFile(ctx, ref)
+		}
+	}
+	return nil, serr
+}
+
+func (v BasicVeneer) renderFile(ctx context.Context, ref string) (*declcfg.DeclarativeConfig, error) {
+	var cfg = &declcfg.DeclarativeConfig{}
+
+	// xform any relative to absolute paths
+	abspath, err := filepath.Abs(ref)
+	if err != nil {
+		return nil, err
+	}
+	// xform to break apart dir/file elements
+	rpath := filepath.Dir(abspath)
+	fname := filepath.Base(abspath)
+
+	root := os.DirFS(rpath)
+	file, err := root.Open(fname)
+	if err != nil {
+		return nil, err
+	}
+
+	defer file.Close()
+
+	if cfg, err = declcfg.LoadFile(root, file, ref); err != nil {
+		return cfg, err
+	}
+
+	var outb = cfg.Bundles[:0] // allocate based on max size of input, but empty slice
+	var r action.Render
+	// populate registry, incl any flags from CLI
+	if v.Registry != nil {
+		r.Registry = v.Registry
+	}
+
+	for _, b := range cfg.Bundles {
+		if isBundleVeneer(&b) {
+			r.Refs = []string{b.Image}
+			contributor, err := r.Run(ctx)
+			if err != nil {
+				// abort processing
+				return cfg, err
+			} else {
+				outb = append(outb, contributor.Bundles...)
+			}
+		} else {
+			outb = append(outb, b)
+		}
+	}
+
+	cfg.Bundles = outb
+	return cfg, nil
+}
+
+// identify loaded partial Bundle data from YAML/JSON veneer source as missing `olm.bundle.object` properties,
+// so `true` if:
+// - no properties
+// - no olm.bundle.object property
+func isBundleVeneer(b *declcfg.Bundle) bool {
```

Comment:
This isn't a good detection mechanism -- in lots of catalogs only the channel head bundles will have `olm.bundle.object` properties.

I think a better detection mechanism would be simply `len(b.Properties) == 0`

Reply:
I did not know.  I've updated to look for empty properties instead.